### ISSUE TITLE
Does apply_buckets_with_interpolation support higher-dimensional inputs?

### ISSUE THREAD

**Issue by cyc** [tensorflow/transform#110]
`tft.apply_buckets_with_interpolation` is a super useful transformation for training neural networks. However, it would be best if I could concatenate all my dense features together and then compute `tft.quantiles` and `tft.apply_buckets_with_interpolation` as one transformation. Is this supported?

Also, to answer the question in the comments of the code (`# TODO(b/127971746): Consider if/how SparseTensor input should be handled.`), my preference would be to have the option to impute with the median (that is, 0.5).

**Comment by KesterTong** [comment]
Generally our approach has been not to impute any value but to leave a missing values in a SparseTensor as missing.  However user code can then fill in missing with 0.5, which I agree is a reasonable value to impute.

**Comment by askerry** [comment]
We've pushed a change to allow `tft.apply_buckets_with_interpolation` to operate on SparseTensors. As Kester mentioned, missing values are left missing, but you can impute with .5 on the output.

**Comment by cyc** [comment]
Great, that is a better solution for the issue of sparse tensor inputs.

Regarding my initial question, I tried hacking something together using `_quantiles_per_key` and `_apply_buckets_with_keys` but the performance was predictably awful. Hopefully you guys can figure out a clever way to get this to work on rank-2 inputs, since for large feature sets the performance is not so great when applying the transform independently for each feature.

**Comment by KesterTong** [comment]
oh that's unfortunate as our plan was to combine _quantiles_per_key and something like _apply_buckets_with_keys.  Which part had poor performance, the quantile computation or the application?

**Comment by cyc** [comment]
@KesterTong it's specifically the quantile computation, and this particular step in the dataflow graph: `AnalyzeAndTransformData/AnalyzeDataset/CacheableCombinePerKeyAccumulate[quantiles_per_key]/CombinePerKey/GroupByKey`. For reference, the dataset consists of ~150 scalar features, which are concatenated together into a rank-1 tensor before being fed into the quantiles_per_key op.

Throughput is about 5-10% of what it is if I just process each feature independently. This somewhat makes sense, though, since we are making the dataflow job do a lot of unnecessary work: our data is already split out by feature, so it is wasteful to merge them all together into a rank-1 tensor and have it partition it back out by feature name.

**Comment by KesterTong** [comment]
Thanks for the info.  That makes sense, as every instance contains every "key" so there is no advantage gained from the shuffle stage.  We are investigating a different approach of using a single combiner that handles many features/dimensions at once.

**Comment by michaelwunder** [comment]
We have now added support for higher dimensions, using the reduce_instance_dims=False option in the quantiles function. This change will reduce the number of Beam calls per feature. Upon computing the boundaries for the input, the buckets will have to be mapped to each feature individually, as is shown in the tft.bucketize mapper.
